File: C++/문제_17281번.cpp
//문제 17281 야구 -> 백트래킹, 시뮬레이션
int v[10];
int n, result;
int hit[51][10];
bool visit[10];
void dfs(int index){
    if(index == 10){
        int score = 0;
        int idx = 1;
        for(int i = 1;i<=n;i++){
            int out = 0;
            int base[4] = {0,0,0,0};
            for(int j = idx;;j++){
                if(j == 10)
                    j = 1;
                //아웃
                if(hit[i][v[j]] == 0){
                    out++;
                    //이닝 종료
                    if(out == 3){
                        idx = j + 1;
                        if(idx == 10)
                            idx = 1;
                        break;
                    }
                }
                //1루타
                else if(hit[i][v[j]] == 1){
                    if(base[3] == 1){
                        score++;
                        base[3] = 0;
                    }
                    for(int k = 2;k>=1;k--){
                        if(base[k] == 1){
                            base[k+1] = 1;
                            base[k] = 0;
                        }
                    }
                    base[1] = 1;
                }
                else if(hit[i][v[j]] == 2){
                    for(int k = 3;k>=2;k--){
                        if(base[k] == 1){
                            base[k] = 0;
                            score++;
                        }
                    }
                    if(base[1]){
                        base[3] = 1;
                        base[1] = 0;
                    }
                    base[2] = 1;
                }
                else if(hit[i][v[j]] == 3){
                    for(int k = 3;k>=1;k--){
                        if(base[k]){
                            score++;
                            base[k] = 0;
                        }
                    }
                    base[3] = 1;
                }
                else{
                    for(int k = 3;k>=1;k--){
                        if(base[k])
                            score++;
                        base[k] = 0;
                    }
                    score++;
                }
            }
        }
        if(result < score)
            result = max(result , score);
        return;
    }
    for(int i = 1;i<=9;i++){
        if(visit[i] == 0){
            visit[i] = 1;
            v[i] = index;
            dfs(index + 1);
            visit[i] = 0;
        }
    }
}
int main(){
    cin.tie(NULL);
    ios_base::sync_with_stdio(false);
    cout.tie(NULL);
    cin >> n;
    for(int i = 1;i<=n;i++){
        for(int j = 1;j<=9;j++){
            cin >> hit[i][j];
        }
    }
    visit[4] = 1;
    v[4] = 1;
    dfs(2);
    cout << result;
}

File: JavaScript/문제_17281번.js
let input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
let n = +input[0];
input.shift();
for (let i = 0; i < input.length; i++) {
    input[i] = input[i].split(' ').map(item => +item);
}
let result = +0;
let visit = new Array(10).fill(false);
//횟수와 타순을 넣는 것
const dfs = (cnt, arr) => {
    if (cnt == 9) {
        let score = +0;
        let number = +0;
        //이닝
        for (let i = 0; i < n; i++) {
            //한 이닝당 아웃 카운트 3
            let out = +0;
            let base = new Array(3).fill(0);//3, 2, 1루
            let j = number;
            for (; ; j++) {
                if (j >= 9)
                    j = 0;
                if (out == 3)
                    break;
                let plus = +0;
                switch (input[i][arr[j]]) {
                    case 0:
                        out++;
                        break;
                    case 1:
                        plus += base[0];
                        base[0] = base[1];
                        base[1] = base[2];
                        base[2] = 1;
                        break;
                    case 2:
                        plus += base[0];
                        plus += base[1];
                        base[0] = base[2];
                        base[1] = 1;
                        base[2] = 0;
                        break;
                    case 3:
                        plus += base[0];
                        plus += base[1];
                        plus += base[2];
                        base[0] = 1;
                        base[1] = 0;
                        base[2] = 0;
                        break;
                    case 4:
                        for (let i = 0; i < 3; i++) {
                            plus += base[i];
                            base[i] = 0;
                        }
                        plus += +1;
                        break;
                }
                score += plus;
            }
            number = j;
        }
        result = Math.max(result, score);
        return;
    }
    //해당 타순이 배치가 되었다면 패스
    if (arr[cnt] != -1) {
        dfs(cnt + 1, arr);
        return;
    }

    //남은 타자
    for (let i = 1; i < 9; i++) {
        //i번째 타자가 배치가 안되었다면
        if (visit[i] == 0) {
            visit[i] = 1;
            arr[cnt] = i;
            dfs(cnt + 1, arr);
            arr[cnt] = -1;
            visit[i] = 0;
        }
    }
}
//4번 타자에 1번을 배치
visit[0] = 1;
dfs(0, [-1, -1, -1, 0, -1, -1, -1, -1, -1]);
console.log(result);